presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -348,14 +395,38 @@ public Type getType()
 
         public void processPage(GroupByIdBlock groupIds, Page page)
         {
-            if (step.isInputRaw()) {
+            if (!this.orderByChannels.isEmpty()) {
+                checkArgument(step == Step.SINGLE && pageIndex.isPresent(), "ORDER BY in aggregations is supported only for single step aggregations");
+                pageIndex.ifPresent(value -> value.addPage(page));
+            }
+            else if (step.isInputRaw()) {
                 aggregation.addInput(groupIds, page);
             }
             else {
                 aggregation.addIntermediate(groupIds, page.getBlock(intermediateChannel));
             }
         }
 
+        public void processOrderBy()
+        {
+            if (this.pageIndex.isPresent()) {
+                checkArgument(groupByHash.isPresent(), "No groupByHash");
+                PagesIndex pageIndex = this.pageIndex.get();
+                pageIndex.sort(orderByChannels, orderings);
+                int currentPosition = 0;
+                while (true) {
+                    pageBuilder.reset();
+                    currentPosition = pageIndex.buildPage(currentPosition, orderByOutputChannels, pageBuilder);
```

Comment:
It is a second place in your PR where you do PagesIndex to Iterator<Page> conversion (this time it's not explicit). 

Implement this once, or better yet, reuse https://github.com/prestodb/presto/pull/8166/commits/9e2bffcda644d7a21547d1e6c9ba0e9288474ead from #8166